src/librustc_driver/driver.rs

```diff
@@ -934,6 +934,7 @@ pub fn phase_3_run_analysis_passes<'tcx, F, R>(sess: &'tcx Session,
     passes.push_pass(MIR_CONST, mir::transform::rustc_peek::SanityCheck);
 
     // What we need to run borrowck etc.
+    passes.push_pass(MIR_CONST, mir::transform::borrow_check::BorrowckMir);
     passes.push_pass(MIR_VALIDATED, mir::transform::qualify_consts::QualifyAndPromoteConstants);
     passes.push_pass(MIR_VALIDATED,
```

Comment:
I don't disagree with your assessment, but I am left somewhat confused as to how things end up in this state. Does this mean that `SimplifyBranches` should be in the `MIR_OPTIMIZED` pipeline instead of the `MIR_VALIDATED` one?